src/spyglass/common/common_interval.py

```diff
@@ -209,6 +225,537 @@ def cleanup(self, dry_run=True):
+
+        if not (
+            interval1[-1][1] + 1 == interval2[0][0]
+            or interval2[0][1] + 1 == interval1[-1][0]
+        ):
+            return Interval(np.concatenate((interval1, interval2), axis=0))
+
+        x = np.array(
+            [
+                [
+                    np.min([interval1[-1][0], interval2[0][0]]),
+                    np.max([interval1[-1][1], interval2[0][1]]),
+                ]
+            ]
+        )
+        return Interval(np.concatenate((interval1[:-1], x), axis=0))
+
+    def union_adjacent_consolidate(self):
+        return Interval(
+            self._expand_1d(reduce(self.union_adjacent_index, self.times))
+        )
+
+    def union(
+        self,
+        other: np.ndarray,
+        min_length: float = 0.0,
+        max_length: float = 1e10,
+    ) -> np.ndarray:
+        """Finds the union (all times in one or both) for two interval lists
+
+        Parameters
+        ----------
+        interval_list1 : np.ndarray
+            The first interval list [start, stop]
+        interval_list2 : np.ndarray
+            The second interval list [start, stop]
+        min_length : float, optional
+            Minimum length of interval for inclusion in output, default 0.0
+        max_length : float, optional
+            Maximum length of interval for inclusion in output, default 1e10
+
+        Returns
+        -------
+        np.ndarray
+            Array of intervals [start, stop]
+        """
+        interval_list1 = self.times
+        interval_list2 = self._extract(other)
+
+        def _parallel_union(interval_list):
+            """Create a parallel list where 1 is start and -1 the end"""
+            interval_list = np.ravel(interval_list)
+            interval_list_start_end = np.ones(interval_list.shape)
+            interval_list_start_end[1::2] = -1
+            return interval_list, interval_list_start_end
+
+        il1, il1_start_end = _parallel_union(interval_list1)
+        il2, il2_start_end = _parallel_union(interval_list2)
+
+        # Concatenate the two lists so we can resort the intervals and apply the
+        # same sorting to the start-end arrays
+        combined_intervals = np.concatenate((il1, il2))
+        ss = np.concatenate((il1_start_end, il2_start_end))
+        sort_ind = np.argsort(combined_intervals)
+        combined_intervals = combined_intervals[sort_ind]
+
+        # a cumulative sum of 1 indicates the beginning of a joint interval; a
+        # cumulative sum of 0 indicates the end
+        union_starts = np.ravel(
+            np.array(np.where(np.cumsum(ss[sort_ind]) == 1))
+        )
+        union_stops = np.ravel(np.array(np.where(np.cumsum(ss[sort_ind]) == 0)))
+        union = [
+            [combined_intervals[start], combined_intervals[stop]]
+            for start, stop in zip(union_starts, union_stops)
+        ]
+
+        return Interval(np.asarray(union))
+
+    def censor(self, timestamps):
```

Comment:
I realize this probably wasn't there previously but it would also be helpful to add types and complete docstrings to all these methods